Data/ByteString/Lazy/ReadNat.hs

```diff
@@ -0,0 +1,200 @@
+#else
+_readNatural kont !bs = loop bs False 1 0 []
+  where
+    -- All done
+    loop s@Empty !valid !cnt !acc ns
+        | valid = let !r = kont (combine cnt acc ns) in Just (r, s)
+        | otherwise = Nothing
+    -- skip empty segment
+    loop (Chunk (BI.BS _ 0) cs) !valid !cnt !acc ns
+        = loop cs valid cnt acc ns
+    loop s@(Chunk c@(BI.BS _ !len) cs) !valid !cnt !acc ns
+        = case natdigits c cnt acc ns of
+            (0, _, _, _) -- nothing usable found, we're done
+                | valid
+                  -> let !r = kont (combine cnt acc ns) in Just (r, s)
+                | otherwise
+                  -> Nothing
+            (!used, !d, !a, ms)
+                | used < len -- input not entirely digits
+                  -> let !r = kont (combine d a ms)
+                         !c' = BU.unsafeDrop used c
+                      in Just (r, Chunk c' cs)
+                | otherwise -- perhaps more digits from the remaining chunks
+                  -> loop cs True d a ms
+#endif
+
+----- Misc functions
+
+-- | Largest decimal digit count that never overflows a Word64
+safeLog :: Word
+safeLog = 19
+
+safeBase :: Natural
+safeBase = 10 ^ safeLog
+
+-- The intermediate representation is a little-endian sequence in base @10^19@,
+-- prefixed by an initial element in base @10^cnt@ for some `cnt` between 1 and
+-- 19.  The final result is obtained by recursive pairwise merging of the tail
+-- followed by a final multiplication by 10^cnt and addition of the head.
+--
+natdigits :: BI.ByteString -- ^ Input chunk
+          -> Word         -- ^ partial digit count
+          -> Word64       -- ^ accumulated element
+          -> [Natural]    -- ^ accumulated MSB elements
+          -> (Int, Word, Word64, [Natural])
+{-# INLINE natdigits #-}
+natdigits (BI.BS !fp !len) = \ cnt acc ns ->
+    BI.accursedUnutterablePerformIO $
+        BI.unsafeWithForeignPtr fp $ \ ptr -> do
+            let end = ptr `plusPtr` len
+            (!cnt', !acc', ns', !next) <- go end cnt acc ns ptr
+            let !n = next `minusPtr` ptr
+            return (n, cnt', acc', ns')
+  where
+    go end = loop
+      where
+        loop :: Word -> Word64 -> [Natural] -> Ptr Word8
+             -> IO (Word, Word64, [Natural], Ptr Word8)
+        loop !cnt !acc ns !ptr
+            | ptr == end = return (cnt, acc, ns, ptr)
+            | otherwise = fromDigit64 . toWord64 <$> peek ptr >>= \ !d ->
+                if d > 9
+                then return (cnt, acc, ns, ptr)
+                else if cnt < safeLog
+                then loop (cnt+1) (10*acc + d) ns $ ptr `plusPtr` 1
+                else loop 1 d (toNatural acc : ns) $ ptr `plusPtr` 1
+
+-- | Merge least-significant head with reduction of of little-endian tail.
+combine :: Word             -- ^ count of digits in head element
+        -> Word64           -- ^ value of head element
+        -> [Natural]        -- ^ tail elements (base @10^'safeBase'@)
```

Comment:
```suggestion
        -> [Natural]        -- ^ tail elements (base @10^'safeLog'@)
```

Reply:
Yes, that's correct.  I could also just say `19`? What do you think?